### PR TITLE
Added min,max and unique to items

### PR DESCRIPTION
**Pull request by sorenbronsted**
Added min,max and unique to items

Trac ticket: https://core.trac.wordpress.org/ticket/48821

### CODE REVIEW

File: src/wp-includes/rest-api.php [hunk 1]
```diff
@@ -1258,12 +1260,48 @@ function rest_validate_value_from_schema( $value, $args, $param = '' ) {
 			return new WP_Error( 'rest_invalid_param', sprintf( __( '%1$s is not of type %2$s.' ), $param, 'array' ) );
 		}
 
+		if ( isset( $args['minItems'] ) ) {
+			if ( $args['minItems'] < 0 ) {
```
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api.php:1264]: We can skip the checks that the schema itself is correct. At some point it'd be a good idea to validate the schema separately, but I don't think we should do that piecemeal and instead think out a solid plan for it. We don't apply that validation in the number constraints for instance.

Comment (sorenbronsted) [reply src/wp-includes/rest-api.php:1264]: Ok

**Comment by sorenbronsted** [comment]
> We should also do research about what the most correct way to check uniqueness would be. `SORT_REGULAR` does an `==` comparison which is of particular importance for [arrays and objects](https://www.php.net/manual/en/language.operators.array.php).
> 
> For instance with the given schema and values, should those be considered unique or not?
> 
> ```json
> {
>   "type": "array",
>   "uniqueItems": true,
>   "items": {
>     "type": "array",
>     "items": {
>       "type": "string"
>     }
>   }
> }
> ```
> 
> ```json
> [
>   [
>     "a",
>     "b",
>     "c"
>   ],
>   [
>     "b",
>     "a",
>     "c"
>   ]
> ]
> ```
> 
> Or what about
> 
> ```json
> {
>   "type": "array",
>   "uniqueItems": true,
>   "items": {
>     "type": [
>       "number",
>       "integer"
>     ]
>   }
> }
> ```
> 
> ```json
> [
>   1.0,
>   1
> ]
> ```
> 
> We should look at the spec, tests and other validators.

I have looked into justinrainbow approach and he uses var_exports and thereby converts everything to strings, but in my opinion he fails on numbers because [1,1.0] is unique according to justinrainbow but I think not, and the spec for uniqueItems is rather vague on what is unique means.
I will use justinrainbow approach but use the flag SORT_NUMERIC when the array type is declared numeric.
Thank you for your patience and feedback on my contributions :-)

**Comment by TimothyBJacobs** [comment]
Could you also take a look at https://github.com/ajv-validator/ajv/?

> Thank you for your patience and feedback on my contributions :-)

No problem at all, thanks for taking on these tickets!

**Comment by sorenbronsted** [comment]
> Could you also take a look at https://github.com/ajv-validator/ajv/?
> 
> > Thank you for your patience and feedback on my contributions :-)
> 
> No problem at all, thanks for taking on these tickets!
 
The testcases I have written are taken are from [official testsuite](https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/master/tests/draft2019-09/uniqueItems.json),  and other implementations uses the same testsuite, so this implementation can the same as others including ajv.

**Comment by TimothyBJacobs** [comment]
Does the official test suite have tests covering the array/object behavior though?

If not, what I think we want is for arrays order should matter. `[ 'a', 'b' ]` is distinct from `[ 'b', 'a' ]`. That means that with an `uniqueItems` keyword the following would pass validation.

```json
[
 [ "a", "b" ],
 [ "b", "a" ]
]
```

```json
{
  "type": "array",
  "uniqueItems": true,
  "items": {
    "type": "array",
    "items": {
      "type": "string"
    }
  }
}
```

For objects, the order _should not_ matter. So the following example should fail validation.

```json
[
  {
    "a": "1",
    "b": "2"
  },
  {
    "b": "2",
    "a": "1"
  }
]
```

```json
{
  "type": "array",
  "uniqueItems": true,
  "items": {
    "type": "object",
    "properties": {
      "a": {
        "type": "string",
      },
      "b": {
        "type": "string"
      }
    }
  }
}
```

This would match the idea that JSON arrays are "ordered" but JSON objects are "unordered".

**Comment by TimothyBJacobs** [comment]
AFAICT that is what ajv does

- https://runkit.com/timothybjacobs/5ef2cd72842f5d001b7f78a5
- https://runkit.com/timothybjacobs/5ef2cd9ccd1fe30013c27062

**Comment by sorenbronsted** [comment]
> Does the official test suite have tests covering the array/object behavior though?

Yes it does. The above link is the official test suite for unique items.

> For objects, the order _should not_ matter. So the following example should fail validation.
> 
> ```json
> [
>   {
>     "a": "1",
>     "b": "2"
>   },
>   {
>     "b": "2",
>     "a": "1"
>   }
> ]
> ```
> 
> ```json
> {
>   "type": "array",
>   "uniqueItems": true,
>   "items": {
>     "type": "object",
>     "properties": {
>       "a": {
>         "type": "string",
>       },
>       "b": {
>         "type": "string"
>       }
>     }
>   }
> }
> ```
> 
> This would match the idea that JSON arrays are "ordered" but JSON objects are "unordered".

My implementation fails on the object case, so I will fix that. Actually I will implement the official test suite for unique items.

**Pull request closed without merging**